### PR TITLE
[ENH] Implement minimal m-separator functionality for enumerating minimal m-separators and also checking if separating set is minimal

### LINKED ISSUE #34: A function for enumerating d-separators in causal graphs
**Is your feature request related to a problem? Please describe.**
In many cases, we are interested in generating/querying d-separators from a causal graph. Polynomial time-delay algorithms exist for such a task and should be added to the `pywhy_graphs.algorithms` submodule.

**Describe the solution you'd like**
Implement https://proceedings.mlr.press/v115/van-der-zander20a.html for ADMG, nx.DiGraph and CPDAG.

This should operate over the subgraphs property in the `MixedEdgeGraph` that operates over different edge's subgraphs.

Additional reference: https://arxiv.org/pdf/1803.00116.pdf

### PR DESCRIPTION
**Pull request by jaron-lee**
Fixes #34 

Changes proposed in this pull request:

- Implements a function to detect minimal m-separators (if any) in a graph

## Before submitting

- [ ] I've read and followed all steps in the [Making a pull request](https://github.com/py-why/pywhy-graphs/blob/main/CONTRIBUTING.md#making-a-pull-request)
    section of the `CONTRIBUTING` docs.
- [ ] I've updated or added any relevant docstrings following the syntax described in the
    [Writing docstrings](https://github.com/py-why/pywhy-graphs/blob/main/CONTRIBUTING.md#writing-docstrings) section of the `CONTRIBUTING` docs.
- [ ] If this PR fixes a bug, I've added a test that will fail without my fix.
- [ ] If this PR adds a new feature, I've added tests that sufficiently cover my new functionality.

## After submitting

- [ ] All GitHub Actions jobs for my pull request have passed.

### CODE REVIEW

File: pywhy_graphs/networkx/algorithms/causal/m_separation.py [hunk 1]
```diff
@@ -179,3 +177,204 @@ def m_separated(
...
+    z_dprime = _bfs_with_marks(aug_G_p, x, z_prime)
+    z = _bfs_with_marks(aug_G_p, y, z_dprime)
+
+    if not m_separated(G, x, y, z, bidirected_edge_name, directed_edge_name, undirected_edge_name):
+        return None
+
+    return z
+
+
+def _bfs_with_marks(G, start_node, check_set):
```
Comment (adam2392) [inline pywhy_graphs/networkx/algorithms/causal/m_separation.py:340]:
```suggestion
# XXX: If networkx makes the corresponding function in `d_separation.py` public, then we can depend on that implementation
def _bfs_with_marks(G, start_node, check_set):
```

**Comment by jaron-lee** [comment]
@adam2392 when you get a chance - I tried implementing minimal m-separators according to the Van der Zander paper. They have an example in figure 5 which I have written tests for. I have a failing test for `minimal_m_separators(G, "X", "Y", i={"Z_1"})` in that graph. 
I think I have implemented the algorithm (FINDMINSEP) correctly as it is written down in the paper, but there seems to be some issues with it generally? For example, when computing `Z''` it's possible that the algorithm would search for a path from `X` to something in `i`, but in the third line of FINDMINSEP all nodes in `i` are removed, so there aren't any paths to `i` at all. I have tried both including and excluding `i` when this happens but couldn't get things to work. Currently I am excluding `i`, but in the failing test case this means that `z` is empty and so we fail to find a minimal m-separator, even though `Z_1` is a valid minimal m-separator.

**Comment by jaron-lee** [comment]
Oh I think I figured it out, there is a missing line that should read:

"Remove from G' all nodes of I" 

I should submit a PR on that paper...

**Comment by adam2392** [comment]
> Oh I think I figured it out, there is a missing line that should read:
> 
> "Remove from G' all nodes of I"
> 
> I should submit a PR on that paper...

If it ends up working, we should document the "change" relative to what's in the paper, so at least our docstring is a pseudo-source of truth :p

File: docs/whats_new/v0.1.rst [hunk 2]
```diff
@@ -26,6 +26,7 @@ Version 0.1
 Changelog
 ---------
 
+- |Feature| Implement minimal m-separator function in :func:`pywhy_graphs.networkx.minimal_m_separators` with a BFS approach, by `Jaron Lee`_ (:pr:`53`)
```
Comment (adam2392) [inline docs/whats_new/v0.1.rst:29]:
```suggestion
- |Feature| Implement minimal m-separator function in :func:`pywhy_graphs.networkx.minimal_m_separator` with a BFS approach, by `Jaron Lee`_ (:pr:`53`)
```

File: pywhy_graphs/networkx/algorithms/causal/m_separation.py [hunk 3]
```diff
@@ -179,3 +178,315 @@ def m_separated(
...
+       Sets in Causal Graphs: Complete Criteria and an Algorithmic Framework,” Artificial
+       Intelligence, vol. 270, pp. 1–40, May 2019, doi: 10.1016/j.artint.2018.12.006.
+
+    """
+    if i is None:
+        i = set()
+    if r is None:
+        r = set(G.nodes())
+    assert i <= z, f"Minimal set {i} should be no larger than proposed separating set {z}"
+    assert z <= r, f"Separating set {z} should be no larger than maximum set {r}"
```
Comment (adam2392) [inline pywhy_graphs/networkx/algorithms/causal/m_separation.py:310]: These have to be `if` statements that result in a e.g. RuntimeError. tldr: assert statements don't always work in Python code if it's compiled into a binary.

Comment (adam2392) [reply pywhy_graphs/networkx/algorithms/causal/m_separation.py:310]: And then a corresponding test for the error cases to complete coverage

**Review by adam2392 — CHANGES_REQUESTED** [review]
I'll do a more thorough check of the algo once we fix the nitpicks. Looking good!

**Comment by jaron-lee** [comment]
@adam2392 I think i basically fixed all the errors, but there is this one CI test failing. Doesn't look related to the code I've written. I vaguely recall we had this on another PR but I don't remember the fix.

**Comment by adam2392** [comment]
> @adam2392 I think i basically fixed all the errors, but there is this one CI test failing. Doesn't look related to the code I've written. I vaguely recall we had this on another PR but I don't remember the fix.

Pushing a fix

File: pywhy_graphs/networkx/algorithms/causal/m_separation.py [hunk 4]
```diff
@@ -179,3 +177,320 @@ def m_separated(
...
+    z : set
+        Set of nodes in ``G`` representing the candidate minimal m-separator of ``x`` and ``y``.
+    i : set
+        Set of nodes which are always included in the found separating set,
+        default is None, which is later set to empty set.
+    r : set
+        Largest set of nodes which may be included in the found separating set,
+        default is None, which is later set to all vertices in ``G``.
+    directed_edge_name : str
+        Name of the directed edge, default is directed.
```
Comment (adam2392) [inline pywhy_graphs/networkx/algorithms/causal/m_separation.py:284]:
```suggestion
        Name of the directed edge, default is 'directed'.
```

File: pywhy_graphs/networkx/algorithms/causal/m_separation.py [hunk 5]
```diff
@@ -179,3 +177,320 @@ def m_separated(
...
+    bidirected_edge_name : str
+        Name of the bidirected edge, default is bidirected.
```
Comment (adam2392) [inline pywhy_graphs/networkx/algorithms/causal/m_separation.py:286]:
```suggestion
        Name of the bidirected edge, default is 'bidirected'.
```

File: pywhy_graphs/networkx/algorithms/causal/m_separation.py [hunk 6]
```diff
@@ -179,3 +177,320 @@ def m_separated(
...
+    undirected_edge_name : str
+        Name of the undirected edge, default is undirected.
```
Comment (adam2392) [inline pywhy_graphs/networkx/algorithms/causal/m_separation.py:288]:
```suggestion
        Name of the undirected edge, default is 'undirected'.
```

File: pywhy_graphs/networkx/algorithms/causal/m_separation.py [hunk 7]
```diff
@@ -179,3 +177,320 @@ def m_separated(
...
+
+    This implements the algorithm TESTMINSEP presented in [1]_ for ancestral mixed graphs, but has
+    been tested to work with non-ancestral graphs.
+
+    This implementation differs from the specification of TESTMINSEP in [1]_ in that we replace
+    the check in line 6 with :math:`z - i \\neq R_x`, and similarly in line 8 with
+    :math:`z - i \\neq R_y`. This change was deemed necessary because in line 4 all nodes in ``i``
+    are removed, so there cannot be paths from ``x`` to ``i`` in the augmented moralized graph,
+    causing the algorithm to return false when a minimal m-separator existed.
+
```
Comment (adam2392) [inline pywhy_graphs/networkx/algorithms/causal/m_separation.py:265]:
```suggestion
```

**Review by adam2392 — APPROVED** [review]
LGTM minus some doc fixes.

**Comment by adam2392** [comment]
https://output.circle-artifacts.com/output/job/622f3036-7e58-4476-9a6a-f9aa8cfa376c/artifacts/0/dev/api.html#networkx-experimental-functionality

**Comment by adam2392** [comment]
Will merge this once CIs give green light

**Comment by adam2392** [comment]
Thanks @jaron-lee !

**Pull request merged**